applications/sdp/mspi/src/hrt/hrt.c

```diff
@@ -6,183 +6,102 @@
-			xfer_ll_params.data_to_send[i] =
-				xfer_ll_params.data_to_send[i]
-				<< (MAX_WORD_SIZE - xfer_ll_params.word_size);
-		}
+	if (xfer_ll_params.ce_polarity == MSPI_CE_ACTIVE_LOW) {
+		out = BIT_SET_VALUE(out, xfer_ll_params.ce_vio, VPRCSR_NORDIC_OUT_LOW);
+	} else {
+		out = BIT_SET_VALUE(out, xfer_ll_params.ce_vio, VPRCSR_NORDIC_OUT_HIGH);
 	}
-
-	/* Counter settings */
-	nrf_vpr_csr_vtim_count_mode_set(0, NRF_VPR_CSR_VTIM_COUNT_RELOAD);
-	nrf_vpr_csr_vtim_simple_counter_top_set(0, xfer_ll_params.counter_top);
-
-	/* Set number of shifts before OUTB needs to be updated.
-	 * First shift needs to be increased by 1.
-	 */
-	nrf_vpr_csr_vio_shift_cnt_out_set(xfer_ll_params.word_size);
-	nrf_vpr_csr_vio_shift_cnt_out_buffered_set(xfer_ll_params.word_size - 1);
-
-	/* Enable CS */
-	out = nrf_vpr_csr_vio_out_get();
-	out &= ~PIN_OUT_HIGH_MASK(VIO(NRFE_MSPI_CS0_PIN_NUMBER));
-	out |= xfer_ll_params.ce_enable_state ? PIN_OUT_HIGH_MASK(VIO(NRFE_MSPI_CS0_PIN_NUMBER))
-					      : PIN_OUT_LOW_MASK(VIO(NRFE_MSPI_CS0_PIN_NUMBER));
 	nrf_vpr_csr_vio_out_set(out);
 
-	/* Start counter */
-	nrf_vpr_csr_vtim_simple_counter_set(0, CLK_FIRST_CYCLE_MULTIPLICATOR *
-						       xfer_ll_params.counter_top);
+	for (uint32_t i = 0; i < xfer_ll_params.words; i++) {
+		switch (xfer_ll_params.words - i) {
+		case 1: /* Last transfer */
+			nrf_vpr_csr_vio_shift_cnt_out_buffered_set(xfer_ll_params.last_word_clocks -
+								   1);
+			switch (xfer_ll_params.bit_order) {
+			case HRT_BO_NORMAL:
+				nrf_vpr_csr_vio_out_buffered_set(xfer_ll_params.last_word);
+				break;
+			case HRT_BO_REVERSED_BYTE:
+				nrf_vpr_csr_vio_out_buffered_reversed_byte_set(
+					xfer_ll_params.last_word);
+				break;
+			case HRT_BO_REVERSED_WORD:
+				nrf_vpr_csr_vio_out_buffered_reversed_word_set(
+					xfer_ll_params.last_word);
+				break;
+			}
+			break;
+		case 2: /* Last but one transfer.*/
+			nrf_vpr_csr_vio_shift_cnt_out_buffered_set(
+				xfer_ll_params.penultimate_word_clocks - 1);
+
+			/* Intentional fallthrough. */
+		default:
+			switch (xfer_ll_params.bit_order) {
+			case HRT_BO_NORMAL:
+				nrf_vpr_csr_vio_out_buffered_set(
+					((uint32_t *)xfer_ll_params.data)[i]);
+				break;
+			case HRT_BO_REVERSED_BYTE:
+				nrf_vpr_csr_vio_out_buffered_reversed_byte_set(
+					((uint32_t *)xfer_ll_params.data)[i]);
+				break;
+			case HRT_BO_REVERSED_WORD:
+				nrf_vpr_csr_vio_out_buffered_reversed_word_set(
+					((uint32_t *)xfer_ll_params.data)[i]);
+				break;
+			}
+		}
 
-	/* Send data */
-	for (uint8_t i = 0; i < xfer_ll_params.data_len; i++) {
-		nrf_vpr_csr_vio_out_buffered_reversed_byte_set(xfer_ll_params.data_to_send[i]);
+		if (i == 0) {
+			/* Start counter */
+			nrf_vpr_csr_vtim_simple_counter_set(0,
+							    xfer_ll_params.counter_initial_value);
+			/* Wait for first clock cycle */
+			shift_count = nrf_vpr_csr_vio_shift_cnt_out_get();
+			while (nrf_vpr_csr_vio_shift_cnt_out_get() == shift_count) {
```

Comment:
In case when second word is not 32 bit, buffered shift count has to be changed while first word is being transfered, otherwise first word will be sent too short. So after starting timer for dummy cycles time, processor has to wait for first CNT0 event before changing settings of shift counter.

Reply:
I am not sure if this is a good solution, I am afraid it will break easily when we change something in hrt code due to timings.
Have you considered sending data byte by byte instead of by 32 bits, as I suggested here?: https://github.com/nrfconnect/sdk-nrf/pull/19077#discussion_r1861186053
Maybe it would be a better solution since sending by 32 bits is causing so much trouble?